File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/protocol_sdk/impl/DeviceLogImpl.swift
//
//  DeviceLogImpl.swift
//  protocol_channel
//
//  Created by hc on 2023/9/21.
//

import Foundation

/// 日志整体进度 0-100
public typealias BlockLogProgress = (_ progress: Double) -> Void

private var _deviceLog: DeviceLog? {
    return SwiftProtocolChannelPlugin.shared.deviceLog
}

private var _delegate: DeviceLogDelegateImpl {
    return DeviceLogDelegateImpl.shared
}

class DeviceLogImpl: IDODeviceLogInterface {
    
    var logDirPath: String {
       return _delegate.logDirPath
    }
    
    var getLogIng: Bool {
        return _delegate.logStatus;
    }
    
    func startGet(types: [IDODeviceLogTypeClass], timeOut: Int, progress: @escaping (Double) -> Void, completion: @escaping (Bool) -> Void) {
        let logTypes = types.map { $0.deviceLogType.rawValue.int64 }
        DeviceLogDelegateImpl.shared.callbackProgress = progress;
        _runOnMainThread {
            _deviceLog?.startGet(types: logTypes, timeOut: timeOut.int64, completion: completion)
        }
    }
    
    func cancel(completion: @escaping () -> Void) {
        _runOnMainThread {
            _deviceLog?.cancel { }
        }
    }
    
}

/// 日志类型
@objc
public enum IDODeviceLogType: Int {
    case initial = 0
    /// 1: 旧的重启日志
    case reboot = 1
    /// 2: 通用日志
    case general = 2
    /// 3: 复位日志
    case reset = 3
    /// 4: 硬件日志
    case hardware = 4
    /// 5: 算法日志
    case algorithm = 5
    /// 6: 新重启日志
    case restart = 6
    /// 7:电池日志
    case battery = 7
    /// 8: 过热日志
    case heat = 8
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/DeviceLogImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_device_log.DeviceLog
import com.idosmart.protocol_channel.plugin
import com.idosmart.protocol_sdk.DeviceLogInterface
import com.idosmart.enums.*
import com.idosmart.protocol_channel.innerRunOnMainThread
import com.idosmart.protocol_sdk.BlockLogProgress

internal class DeviceLogImpl: DeviceLogInterface {

    private fun deviceLog(): DeviceLog? {
        return plugin.deviceLog()
    }

    override val getLogIng: Boolean
        get() = DeviceLogDelegateImpl.instance().logStatus
    override val logDirPath: String
        get() = DeviceLogDelegateImpl.instance().logDirPath

    override fun startGet(
        types: List<IDODeviceLogType>,
        timeOut: Int,
        progress: BlockLogProgress,
        completion: (result: Boolean) -> Unit
    ) {
        var logTypes = mutableListOf<Long>()
        types.map {
            logTypes.add(it.raw.toLong())
        }
        DeviceLogDelegateImpl.instance().callbackProgress = progress
        innerRunOnMainThread {
            deviceLog()?.startGet(logTypes, timeOut.toLong(), completion)
        }
    }

    override fun cancel(completion: () -> Unit) {
        innerRunOnMainThread {
            deviceLog()?.cancel(completion)
        }
    }

}